File: AlgorithmLcr/Try1-Completely unknown single port fitting/rlc_id/fit_engine_a.py
"""Engine A: complex-domain weighted least squares over a topology library.

DESIGN.md section 5:
  * residuals stack real/imaginary parts of w_k * (z_hat_k - Z(jw_k; theta))
    with w_k = 1/sigma_k or, by default, w_k = 1/|z_hat_k| (section 5.1, D4);
  * Jacobians by the complex-step method with h = 1e-20 on the log10
    parameters (section 5.2) -- Z is analytic in theta, so the derivative is
    exact to machine precision with no subtractive cancellation;
  * scipy least_squares TRF with per-kind bounds (section 4.3);
  * multi-start: Latin-hypercube starts plus asymptotic heuristic starts
    (section 5.4), organized as a two-stage coarse -> refine funnel (F4);
  * goodness of fit: AICc with n_obs = 2M and K = p + 1 (section 5.5).
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
from scipy.optimize import least_squares
from scipy.stats import qmc

from .circuits import (KIND_BOUNDS, DCR_BOUNDS, Tree, bounds, canonical,
                       evaluate, evaluate_jac, param_kinds, n_params)

COMPLEX_STEP_H = 1e-20


# ---------------------------------------------------------------------------
# residuals / jacobian / metrics
# ---------------------------------------------------------------------------

def default_weights(z: np.ndarray) -> np.ndarray:
    """Default relative-error weights w_k = 1/|z_hat_k| (model A3)."""
    return 1.0 / np.abs(z)


def residual_vector(tree: Tree, theta, s: np.ndarray, z: np.ndarray,
                    w: np.ndarray) -> np.ndarray:
    """Interleaved [Re r_1, Im r_1, Re r_2, ...] weighted residual."""
    r = w * (z - evaluate(tree, theta, s))
    out = np.empty(2 * len(z))
    out[0::2] = r.real
    out[1::2] = r.imag
    return out


def jacobian_cs(tree: Tree, theta, s: np.ndarray, w: np.ndarray) -> np.ndarray:
    """Exact Jacobian of the weighted residual wrt log10 parameters.

    Uses forward-mode AD (circuits.evaluate_jac).  NOTE: DESIGN.md section 5.2
    specified a complex-step derivative, but that method is invalid here --
    Z is complex-valued at real theta, so Im[Z(theta+i h)]/h is dominated by
    Im[Z(theta)]/h rather than the derivative.  Forward AD is exact to
    machine precision at the cost of a single tree evaluation."""
    _, J = evaluate_jac(tree, theta, s)          # J[i, k] = dZ_k/dtheta_i
    p = J.shape[0]
    out = np.empty((2 * len(s), p))
    dr = -w[None, :] * J                          # d(weighted residual)/dtheta
    out[0::2, :] = dr.real.T
    out[1::2, :] = dr.imag.T
    return out


def rss_of(residual: np.ndarray) -> float:
    return float(residual @ residual)


def aicc(rss: float, n_obs: int, p: int) -> float:
    """Corrected AIC (section 5.5); n_obs = 2M real equations, K = p + 1."""
    k = p + 1
    rss = max(rss, 1e-300)
    denom = max(n_obs - k - 1, 1)
    return n_obs * np.log(rss / n_obs) + 2 * k + 2 * k * (k + 1) / denom


def fit_metrics(z: np.ndarray, zfit: np.ndarray) -> tuple[float, float]:
    """(weighted) relative RMSE and max relative error (section 8.1)."""
    rel = np.abs((z - zfit) / z)
    return float(np.sqrt(np.mean(rel**2))), float(np.max(rel))


# ---------------------------------------------------------------------------
# candidate container
# ---------------------------------------------------------------------------

@dataclass
class Candidate:
    """One fitted topology: structure + log10 parameters + quality metrics."""

    tree: Tree
    theta: np.ndarray
    rss: float
    aicc_val: float
    wrmse: float
    max_rel_err: float
    engine: str = "A"
    note: str = ""
    skipped: bool = False

    @property
    def n_params(self) -> int:
        return len(self.theta)

    @property
    def canonical(self) -> str:
        return canonical(self.tree)

    @property
    def values(self) -> np.ndarray:
        return np.power(10.0, self.theta)


# ---------------------------------------------------------------------------
# heuristic starting points (section 5.4)
# ---------------------------------------------------------------------------

@dataclass
class StartHints:
    """Data-driven magnitudes used to seed the optimizer (from pruning.py)."""

    r_level: float
    l_est: float
    c_est: float
    w_res: float | None = None  # angular frequency of a |Z| extremum, if any
    r_peak: float | None = None  # max|Z|: parallel-resonance R level


def _clip_kind(kind: str, value: float) -> float:
    lo, hi = KIND_BOUNDS[kind]
    return float(np.clip(value, 10.0**lo, 10.0**hi))


def heuristic_starts(tree: Tree, hints: StartHints | None) -> list[np.ndarray]:
    """Asymptotic heuristic starts for this topology (section 5.4).

    start 0: plain asymptotic magnitudes (R from the flat level, L from the
             high-frequency slope, C from the low-frequency slope; the DCR
             parameter of an L device starts at the geometric middle of its
             search box -- no data-driven estimate is robust enough);
    start 1: resonance-informed (only when the data show an interior |Z|
             extremum and the topology contains both L and C): L*C = 1/w0^2.
    """
    pks = param_kinds(tree)
    lb, ub = bounds(tree)
    mid = 0.5 * (lb + ub)
    rd_mid = 0.5 * (DCR_BOUNDS[0] + DCR_BOUNDS[1])

    def base_estimate(pk: str, i: int) -> float:
        if hints is None:
            return rd_mid if pk == "Rd" else mid[i]
        if pk == "R":
            return np.log10(_clip_kind("R", hints.r_level))
        if pk == "L":
            return np.log10(_clip_kind("L", hints.l_est))
        if pk == "Rd":
            return rd_mid
        return np.log10(_clip_kind("C", hints.c_est))

    starts = [np.array([base_estimate(pk, i) for i, pk in enumerate(pks)])]

    if hints is not None and hints.w_res is not None and "L" in pks and "C" in pks:
        th = starts[0].copy()
        l_first = np.log10(_clip_kind("L", hints.l_est))
        for i, pk in enumerate(pks):
            if pk == "L":
                th[i] = l_first
        for i, pk in enumerate(pks):
            if pk == "C":
                lc = 1.0 / hints.w_res**2
                th[i] = np.log10(_clip_kind("C", lc / 10.0**l_first))
        # parallel resonance: R is at the peak of |Z|, not the band median
        if hints.r_peak is not None:
            for i, pk in enumerate(pks):
                if pk == "R":
                    th[i] = np.log10(_clip_kind("R", hints.r_peak))
        starts.append(np.clip(th, lb, ub))
    return [np.clip(st, lb, ub) for st in starts]


def lhs_starts(n: int, lb: np.ndarray, ub: np.ndarray,
               rng: np.random.Generator) -> list[np.ndarray]:
    """Latin-hypercube starts over the log10 bound box."""
    if n <= 0:
        return []
    d = len(lb)
    unit = qmc.LatinHypercube(d=d, seed=int(rng.integers(0, 2**31 - 1))).random(n)
    return [lb + u * (ub - lb) for u in unit]


# ---------------------------------------------------------------------------
# single-topology fit
# ---------------------------------------------------------------------------

def fit_topology(tree: Tree, s: np.ndarray, z: np.ndarray, w: np.ndarray,
                 starts: list[np.ndarray], max_nfev: int | None = None,
                 tol: float = 1e-11) -> Candidate | None:
    """Fit one topology from several starts; return the best candidate."""
    lb, ub = bounds(tree)
    p = len(lb)
    best: Candidate | None = None
    for x0 in starts:
        x0 = np.clip(np.asarray(x0, dtype=float), lb + 1e-12, ub - 1e-12)
        try:
            res = least_squares(
                lambda th: residual_vector(tree, th, s, z, w),
                x0,
                jac=lambda th: jacobian_cs(tree, th, s, w),
                bounds=(lb, ub),
                method="trf",
                loss="linear",
                max_nfev=max_nfev,
                xtol=tol,
                ftol=tol,
                gtol=tol,
            )
        except Exception:
            continue
        rss = rss_of(res.fun)
        if best is None or rss < best.rss:
            zfit = evaluate(tree, res.x, s)
            wrmse, emax = fit_metrics(z, zfit)
            best = Candidate(tree=tree, theta=res.x, rss=rss,
                             aicc_val=aicc(rss, 2 * len(z), p),
                             wrmse=wrmse, max_rel_err=emax)
    return best


# ---------------------------------------------------------------------------
# library fit with two-stage funnel (F4)
# ---------------------------------------------------------------------------

@dataclass
class EngineAConfig:
    n_starts_coarse: int = 3
    n_starts_refine: int = 10
    refine_fraction: float = 0.2
    max_nfev_coarse: int = 40
    max_nfev_refine: int | None = None
    tol_coarse: float = 1e-6
    tol_refine: float = 1e-11
    seed: int = 0


def fit_library(trees: list[Tree], s: np.ndarray, z: np.ndarray,
                w: np.ndarray, config: EngineAConfig | None = None,
                hints: StartHints | None = None,
                extra_starts: dict[str, list[np.ndarray]] | None = None,
                ) -> list[Candidate]:
    """Fit every topology in ``trees`` (two-stage coarse -> refine, F4).

    ``extra_starts`` maps a canonical topology string to additional starting
    vectors (used to inject engine-B Foster solutions as starts, which also
    guarantees such topologies survive the coarse stage).
    """
    cfg = config or EngineAConfig()
    rng = np.random.default_rng(cfg.seed)
    extra_starts = extra_starts or {}

    # ---- stage 1: coarse pass over all topologies with few starts ----------
    coarse: list[Candidate] = []
    for tree in trees:
        lb, ub = bounds(tree)
        starts = heuristic_starts(tree, hints)
        starts += list(extra_starts.get(canonical(tree), []))
        n_lhs = max(cfg.n_starts_coarse - len(starts), 1)
        starts += lhs_starts(n_lhs, lb, ub, rng)
        cand = fit_topology(tree, s, z, w, starts,
                            max_nfev=cfg.max_nfev_coarse, tol=cfg.tol_coarse)
        if cand is not None:
            coarse.append(cand)
    if not coarse:
        return []

    # ---- stage 2: refine the best fraction with the full start set ---------
    coarse.sort(key=lambda c: c.rss)
    n_ref = max(1, int(np.ceil(cfg.refine_fraction * len(coarse))))
    # topologies carrying engine-B starts always advance
    front = {c.canonical for c in coarse[:n_ref]} | set(extra_starts)
    refined: list[Candidate] = []
    for cand in coarse:
        if cand.canonical not in front:
            refined.append(cand)
            continue
        tree = cand.tree
        lb, ub = bounds(tree)
        starts = [cand.theta]
        starts += heuristic_starts(tree, hints)
        starts += list(extra_starts.get(canonical(tree), []))
        n_lhs = max(cfg.n_starts_refine - len(starts), 2)
        starts += lhs_starts(n_lhs, lb, ub, rng)
        best = fit_topology(tree, s, z, w, starts,
                            max_nfev=cfg.max_nfev_refine, tol=cfg.tol_refine)
        refined.append(best if best is not None else cand)
    refined.sort(key=lambda c: c.aicc_val)
    return refined

File: AlgorithmLcr/Try1-Completely unknown single port fitting/cppversion/src/fit_engine_a.cpp
#include "fit_engine_a.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <set>

namespace rlc {

std::vector<double> defaultWeights(const std::vector<Complex>& z) {
    std::vector<double> w(z.size());
    for (size_t k = 0; k < z.size(); ++k) w[k] = 1.0 / std::abs(z[k]);
    return w;
}

std::vector<double> residualVector(const TreePtr& tree, const std::vector<double>& theta,
                                   const std::vector<Complex>& s,
                                   const std::vector<Complex>& z,
                                   const std::vector<double>& w) {
    const size_t m = z.size();
    std::vector<Complex> zfit(m);
    evalTheta(tree, theta, s.data(), m, zfit.data());
    std::vector<double> out(2 * m);
    for (size_t k = 0; k < m; ++k) {
        Complex r = w[k] * (z[k] - zfit[k]);
        out[2 * k] = r.real();
        out[2 * k + 1] = r.imag();
    }
    return out;
}

std::vector<double> jacobianCs(const TreePtr& tree, const std::vector<double>& theta,
                               const std::vector<Complex>& s,
                               const std::vector<double>& w) {
    const size_t m = s.size();
    const size_t p = theta.size();
    std::vector<Complex> z(m);
    std::vector<Complex> J(p * m);
    evalJac(tree, theta, s.data(), m, z.data(), J.data());
    std::vector<double> out(2 * m * p);
    for (size_t k = 0; k < m; ++k) {
        for (size_t i = 0; i < p; ++i) {
            Complex dr = -(double)w[k] * J[i * m + k];
            out[(2 * k) * p + i] = dr.real();
            out[(2 * k + 1) * p + i] = dr.imag();
        }
    }
    return out;
}

double rssOf(const std::vector<double>& residual) {
    double s = 0.0;
    for (double v : residual) s += v * v;
    return s;
}

double aicc(double rss, int n_obs, int p) {
    int k = p + 1;
    rss = std::max(rss, 1e-300);
    double denom = std::max((double)(n_obs - k - 1), 1.0);
    return n_obs * std::log(rss / n_obs) + 2 * k + 2 * k * (k + 1) / denom;
}

std::pair<double, double> fitMetrics(const std::vector<Complex>& z,
                                     const std::vector<Complex>& zfit) {
    double s2 = 0.0, mx = 0.0;
    for (size_t k = 0; k < z.size(); ++k) {
        double rel = std::abs((z[k] - zfit[k]) / z[k]);
        s2 += rel * rel;
        mx = std::max(mx, rel);
    }
    return {std::sqrt(s2 / (double)z.size()), mx};
}

std::vector<double> Candidate::values() const {
    std::vector<double> v(theta.size());
    for (size_t i = 0; i < theta.size(); ++i) v[i] = std::pow(10.0, theta[i]);
    return v;
}

// R7: outlier-robust refit of one candidate (single IRLS pass).  A few wild
// measurement points bend a plain least-squares fit away from the inlier
// core; the robust scale is estimated per axis of the COMPLEX relative
// residual (1.4826 x median|x| = sigma for a Gaussian axis), points whose
// residual magnitude exceeds 5 sigma are quadratically downweighted, and the
// candidate is re-fit under the downweighted weights from its own solution.
// Returns true when the parameters changed; metrics (rss/wrmse/aicc) are
// recomputed with the ORIGINAL weights so candidates stay comparable.
bool robustRefitCandidate(Candidate& c, const std::vector<Complex>& s,
                          const std::vector<Complex>& z,
                          const std::vector<double>& w) {
    const size_t m = z.size();
    if (m < 8) return false;
    // model at the current parameters
    std::vector<Complex> zfit(m);
    evalTheta(c.tree, c.theta, s.data(), m, zfit.data());
    std::vector<double> magRe(m), magIm(m), mag(m);
    for (size_t k = 0; k < m; ++k) {
        Complex rr = (z[k] - zfit[k]) / z[k];
        magRe[k] = std::fabs(rr.real());
        magIm[k] = std::fabs(rr.imag());
        mag[k] = std::abs(rr);
    }
    auto medOf = [](std::vector<double> v) {
        std::sort(v.begin(), v.end());
        return v[v.size() / 2];
    };
    double sigAxis = std::max(1.4826 * std::max(medOf(magRe), medOf(magIm)), 1e-9);
    const double kCut = 5.0, kIn = 2.5;
    std::vector<double> wR = w;
    bool anyOut = false;
    int nInlier = 0, nOut = 0;
    for (size_t k = 0; k < m; ++k) {
        double r = mag[k] / sigAxis;
        if (r > kCut) {
            double f = kCut / r;
            wR[k] *= f * f;
            anyOut = true;
            ++nOut;
        } else if (r < kIn) {
            ++nInlier;
        }
    }
    if (!anyOut || nOut > (int)m / 3 || nInlier * 2 < (int)m) return false;

    // acceptance is judged on the ROBUST objective: the honest (inlier)
    // parameters necessarily have a HIGHER plain-weighted rss than the bent
    // least-squares solution, because the plain optimum minimises exactly
    // that.  Comparing plain rss would always reject the rescue.
    auto robustRss = [&](const std::vector<double>& theta) {
        std::vector<Complex> zf(m);
        evalTheta(c.tree, theta, s.data(), m, zf.data());
        double r = 0.0;
        for (size_t k = 0; k < m; ++k) {
            Complex e = wR[k] * (z[k] - zf[k]);
            r += e.real() * e.real() + e.imag() * e.imag();
        }
        return r;
    };
    double rssInc = robustRss(c.theta);

    std::vector<double> lb, ub;
    thetaBounds(c.tree, lb, ub);
    LMOpts opts;
    opts.maxNfev = 3000;
    opts.ftol = opts.xtol = opts.gtol = 1e-12;
    auto residual = [&](const std::vector<double>& th, std::vector<double>& out) {
        out = residualVector(c.tree, th, s, z, wR);
    };
    auto jac = [&](const std::vector<double>& th, std::vector<double>& out) {
        out = jacobianCs(c.tree, th, s, wR);
    };
    LMOut res = lmFit(residual, jac, c.theta, lb, ub, opts);
    double rssR = rssOf(res.residual);
    if (!std::isfinite(rssR) || rssR >= rssInc) return false;

    // metrics with the original weights at the robust parameters
    std::vector<double> resPlain = residualVector(c.tree, res.x, s, z, w);
    double rss = rssOf(resPlain);
    std::vector<Complex> zfit2(m);
    evalTheta(c.tree, res.x, s.data(), m, zfit2.data());
    auto [wrmse, mre] = fitMetrics(z, zfit2);
    c.theta = std::move(res.x);
    c.rss = rss;
    c.aiccVal = aicc(rss, (int)(2 * m), (int)c.theta.size());
    c.wrmse = wrmse;
    c.maxRelErr = mre;
    return true;
}

std::vector<std::vector<double>> heuristicStarts(const TreePtr& tree,
                                                 const StartHints* hints) {
    std::vector<char> kinds = paramKinds(tree);  // 'D' = Rd parameter of L
    std::vector<double> lb, ub;
    thetaBounds(tree, lb, ub);
    const size_t p = kinds.size();
    std::vector<double> mid(p);
    for (size_t i = 0; i < p; ++i) mid[i] = 0.5 * (lb[i] + ub[i]);
    auto dcrBounds = kindBounds('D');
    const double rdMid = 0.5 * (dcrBounds.first + dcrBounds.second);

    auto baseEstimate = [&](char kind, size_t i) -> double {
        if (hints == nullptr) return kind == 'D' ? rdMid : mid[i];
        if (kind == 'R') return std::log10(clipKind('R', hints->rLevel));
        if (kind == 'L') return std::log10(clipKind('L', hints->lEst));
        if (kind == 'D') return rdMid;  // no robust data-driven Rd estimate
        return std::log10(clipKind('C', hints->cEst));
    };

    std::vector<double> s0(p);
    for (size_t i = 0; i < p; ++i) s0[i] = baseEstimate(kinds[i], i);
    std::vector<std::vector<double>> starts{s0};

    bool hasL = false, hasC = false;
    for (char k : kinds) {
        if (k == 'L') hasL = true;
        if (k == 'C') hasC = true;
    }
    if (hints != nullptr && hints->hasWRes && hasL && hasC) {
        std::vector<double> th = s0;
        double lFirst = std::log10(clipKind('L', hints->lEst));
        for (size_t i = 0; i < p; ++i)
            if (kinds[i] == 'L') th[i] = lFirst;
        for (size_t i = 0; i < p; ++i) {
            if (kinds[i] == 'C') {
                double lc = 1.0 / (hints->wRes * hints->wRes);
                th[i] = std::log10(clipKind('C', lc / std::pow(10.0, lFirst)));
            }
        }
        if (hints->hasRPeak) {
            for (size_t i = 0; i < p; ++i)
                if (kinds[i] == 'R') th[i] = std::log10(clipKind('R', hints->rPeak));
        }
        for (size_t i = 0; i < p; ++i)
            th[i] = std::min(std::max(th[i], lb[i]), ub[i]);
        starts.push_back(std::move(th));
    }
    for (auto& st : starts) {
        for (size_t i = 0; i < p; ++i)
            st[i] = std::min(std::max(st[i], lb[i]), ub[i]);
    }
    return starts;
}

// ---------------------------------------------------------------------------
// box-constrained Levenberg-Marquardt
// ---------------------------------------------------------------------------

LMOut lmFit(std::function<void(const std::vector<double>&, std::vector<double>&)> residual,
            std::function<void(const std::vector<double>&, std::vector<double>&)> jac,
            std::vector<double> x0, const std::vector<double>& lb,
            const std::vector<double>& ub, const LMOpts& opts) {
    const size_t p = x0.size();
    // Python: x0 = clip(x0, lb + 1e-12, ub - 1e-12)
    for (size_t i = 0; i < p; ++i) {
        x0[i] = std::min(std::max(x0[i], lb[i] + 1e-12), ub[i] - 1e-12);
        if (!std::isfinite(x0[i])) x0[i] = 0.5 * (lb[i] + ub[i]);
    }
    int maxNfev = opts.maxNfev > 0 ? opts.maxNfev : (int)(100 * p);
    const int m = (int)p;

    std::vector<double> r;
    residual(x0, r);
    double rss = rssOf(r);
    std::vector<double> x = x0;

    std::vector<double> J;
    std::vector<double> g(p), H(p * p), dx(p), rTry;
    std::vector<double> Hsolved;

    double lambda = -1.0;  // initialised from the first Jacobian
    double nu = 2.0;
    int nfev = 1;
    const int maxIter = 200 * (int)p + 200;

    for (int iter = 0; iter < maxIter; ++iter) {
        // convergence: projected gradient
        jac(x, J);
        const int nrows = (int)J.size() / (int)p;
        std::fill(g.begin(), g.end(), 0.0);
        std::fill(H.begin(), H.end(), 0.0);
        for (int k = 0; k < nrows; ++k) {
            const double* row = J.data() + (size_t)k * p;
            double rk = r[k];
            for (int i = 0; i < m; ++i) {
                g[i] += row[i] * rk;
                for (int j = i; j < m; ++j) H[(size_t)i * m + j] += row[i] * row[j];
            }
        }
        for (int i = 0; i < m; ++i)
            for (int j = 0; j < i; ++j) H[(size_t)i * m + j] = H[(size_t)j * m + i];

        double ginf = 0.0;
        for (int i = 0; i < m; ++i) ginf = std::max(ginf, std::fabs(g[i]));
        if (ginf <= opts.gtol * std::max(1.0, rss)) break;

        if (lambda < 0.0) {
            double hmax = 0.0;
            for (int i = 0; i < m; ++i) hmax = std::max(hmax, H[(size_t)i * m + i]);
            lambda = 1e-3 * std::max(hmax, 1e-12);
            nu = 2.0;
        }

        bool stepped = false;
        for (int attempt = 0; attempt < 60 && !stepped; ++attempt) {
            // damped normal equations (H + lambda diag(H)) dx = -g
            Hsolved = H;
            for (int i = 0; i < m; ++i) {
                double& d = Hsolved[(size_t)i * m + i];
                d += lambda * std::max(d, 1e-12);
            }
            dx = g;
            for (double& v : dx) v = -v;
            bool ok = solveSPD(Hsolved, m, dx);
            bool finite = ok;
            if (ok) {
                for (double v : dx) {
                    if (!std::isfinite(v)) {
                        finite = false;
                        break;
                    }
                }
            }
            if (!finite) {
                lambda = (lambda > 0.0) ? lambda * 10.0 : 1e-12;
                nu = 2.0;
                continue;
            }

            std::vector<double> xTry(p);
            for (size_t i = 0; i < p; ++i)
                xTry[i] = std::min(std::max(x[i] + dx[i], lb[i]), ub[i]);
            residual(xTry, rTry);
            ++nfev;

            double rssTry = rssOf(rTry);
            // predicted reduction = -(g.dx + dx.H.dx / 2), positive for a
            // descent step
            double q = 0.0;
            for (int i = 0; i < m; ++i) q += g[i] * dx[i];
            for (int i = 0; i < m; ++i)
                for (int j = 0; j < m; ++j)
                    q += 0.5 * dx[i] * H[(size_t)i * m + j] * dx[j];
            double pred = -q;
            double denom = std::max(pred, 1e-300);
            double rho = (rss - rssTry) / denom;
            double rssFinite =
                std::isfinite(rssTry) ? rssTry : std::numeric_limits<double>::infinity();

            if (rho > 0.0 && rssFinite < rss) {
                // accept
                double stepNorm = 0.0, xNorm = 0.0;
                for (size_t i = 0; i < p; ++i) {
                    double d = xTry[i] - x[i];
                    stepNorm += d * d;
                    xNorm += xTry[i] * xTry[i];
                }
                stepNorm = std::sqrt(stepNorm);
                xNorm = std::sqrt(xNorm);
                double actualRed = rss - rssTry;
                double rssOld = rss;
                x = xTry;
                r = rTry;
                rss = rssTry;
                double t = std::max(1.0 / 3.0, 1.0 - std::pow(2.0 * rho - 1.0, 3));
                lambda *= t;
                if (lambda < 1e-18) lambda = 1e-18;
                nu = 2.0;
                stepped = true;
                if (actualRed <= opts.ftol * std::max(rssOld, 1e-300)) break;
                if (stepNorm <= opts.xtol * (opts.xtol + xNorm)) break;
                if (pred <= opts.ftol * std::max(rssOld, 1e-300)) break;
            } else {
                lambda *= nu;
                nu *= 2.0;
                if (lambda > 1e14) break;  // stalled at a bound / flat region
            }
            if (nfev >= maxNfev) break;
        }
        if (!stepped) break;      // no progress possible
        if (nfev >= maxNfev) break;
    }

    LMOut out;
    out.x = std::move(x);
    out.residual = std::move(r);
    out.rss = rss;
    out.nfev = nfev;
    return out;
}

std::optional<Candidate> fitTopology(const TreePtr& tree, const std::vector<Complex>& s,
                                     const std::vector<Complex>& z,
                                     const std::vector<double>& w,
                                     const std::vector<std::vector<double>>& starts,
                                     int maxNfev, double tol) {
    std::vector<double> lb, ub;
    thetaBounds(tree, lb, ub);
    const size_t p = lb.size();
    std::optional<Candidate> best;

    auto residual = [&](const std::vector<double>& th, std::vector<double>& out) {
        out = residualVector(tree, th, s, z, w);
    };
    auto jac = [&](const std::vector<double>& th, std::vector<double>& out) {
        out = jacobianCs(tree, th, s, w);
    };

    for (const auto& st : starts) {
        LMOpts opts;
        opts.maxNfev = maxNfev;
        opts.ftol = opts.xtol = opts.gtol = tol;
        LMOut res;
        try {
            res = lmFit(residual, jac, st, lb, ub, opts);
        } catch (...) {
            continue;
        }
        double rss = rssOf(res.residual);
        if (!best.has_value() || rss < best->rss) {
            std::vector<Complex> zfit(z.size());
            evalTheta(tree, res.x, s.data(), z.size(), zfit.data());
            auto [wrmse, emax] = fitMetrics(z, zfit);
            Candidate c;
            c.tree = tree;
            c.theta = res.x;
            c.rss = rss;
            c.aiccVal = aicc(rss, (int)(2 * z.size()), (int)p);
            c.wrmse = wrmse;
            c.maxRelErr = emax;
            best = std::move(c);
        }
    }
    return best;
}

std::vector<Candidate> fitLibrary(const std::vector<TreePtr>& trees,
                                  const std::vector<Complex>& s,
                                  const std::vector<Complex>& z,
                                  const std::vector<double>& w,
                                  const EngineAConfig& config,
                                  const StartHints* hints,
                                  const std::map<std::string, std::vector<std::vector<double>>>* extraStarts) {
    Rng rng(config.seed);
    std::vector<Candidate> coarse;
    for (const auto& tree : trees) {
        std::vector<double> lb, ub;
        thetaBounds(tree, lb, ub);
        std::vector<std::vector<double>> starts = heuristicStarts(tree, hints);
        if (extraStarts) {
            auto it = extraStarts->find(canonical(tree));
            if (it != extraStarts->end())
                for (const auto& v : it->second) starts.push_back(v);
        }
        int nLhs = std::max(config.nStartsCoarse - (int)starts.size(), 1);
        auto lhs = lhsStarts(nLhs, lb, ub, rng);
        for (auto& v : lhs) starts.push_back(std::move(v));
        auto cand = fitTopology(tree, s, z, w, starts, config.maxNfevCoarse, config.tolCoarse);
        if (cand.has_value()) coarse.push_back(std::move(*cand));
    }
    if (coarse.empty()) return {};

    // stage 2: refine the best fraction with the full start set
    std::stable_sort(coarse.begin(), coarse.end(),
                     [](const Candidate& a, const Candidate& b) { return a.rss < b.rss; });
    int nRef = std::max(1, (int)std::ceil(config.refineFraction * (double)coarse.size()));
    std::set<std::string> front;
    for (int i = 0; i < nRef && i < (int)coarse.size(); ++i)
        front.insert(coarse[i].canonicalStr());
    if (extraStarts)
        for (const auto& kv : *extraStarts) front.insert(kv.first);

    std::vector<Candidate> refined;
    for (auto& cand : coarse) {
        if (front.find(cand.canonicalStr()) == front.end()) {
            refined.push_back(cand);
            continue;
        }
        const TreePtr& tree = cand.tree;
        std::vector<double> lb, ub;
        thetaBounds(tree, lb, ub);
        std::vector<std::vector<double>> starts{cand.theta};
        for (auto& v : heuristicStarts(tree, hints)) starts.push_back(std::move(v));
        if (extraStarts) {
            auto it = extraStarts->find(canonical(tree));
            if (it != extraStarts->end())
                for (const auto& v : it->second) starts.push_back(v);
        }
        int nLhs = std::max(config.nStartsRefine - (int)starts.size(), 2);
        auto lhs = lhsStarts(nLhs, lb, ub, rng);
        for (auto& v : lhs) starts.push_back(std::move(v));
        auto best = fitTopology(tree, s, z, w, starts, config.maxNfevRefine, config.tolRefine);
        refined.push_back(best.has_value() ? std::move(*best) : cand);
    }
    std::stable_sort(refined.begin(), refined.end(), [](const Candidate& a, const Candidate& b) {
        return a.aiccVal < b.aiccVal;
    });
    return refined;
}

}  // namespace rlc
